server/src/main/java/io/crate/types/NumericStorage.java

```diff
@@ -0,0 +1,193 @@
+        Integer precision = numericType.numericPrecision();
+        if (precision == null || numericType.scale() == null) {
+            throw new UnsupportedOperationException(
+                "NUMERIC type requires precision and scale to support storage");
+        } else if (precision <= COMPACT_PRECISION) {
+            return new CompactNumericIndexer(ref);
+        } else {
+            return new FixedNumericIndexer(ref, precision);
+        }
+    }
+
+    private abstract static class BaseNumericIndexer implements ValueIndexer<BigDecimal> {
+
+        protected final Reference ref;
+        protected final String name;
+
+        protected BaseNumericIndexer(Reference ref) {
+            this.ref = ref;
+            this.name = ref.storageIdent();
+        }
+
+        @Override
+        public String storageIdentLeafName() {
+            return ref.storageIdentLeafName();
+        }
+    }
+
+    private static class CompactNumericIndexer extends BaseNumericIndexer {
+
+        private CompactNumericIndexer(Reference ref) {
+            super(ref);
+        }
+
+        @Override
+        public void indexValue(@NotNull BigDecimal value, IndexDocumentBuilder docBuilder) throws IOException {
+            BigInteger unscaled = value.unscaledValue();
+            long longValue = unscaled.longValueExact();
+
+            if (this.ref.indexType() != IndexType.NONE) {
+                docBuilder.addField(new LongPoint(name, longValue));
+            }
+            if (ref.hasDocValues()) {
+                docBuilder.addField(new SortedNumericDocValuesField(name, longValue));
+            } else {
+                docBuilder.addField(new Field(
+                    DocSysColumns.FieldNames.NAME,
+                    name,
+                    DocSysColumns.FieldNames.FIELD_TYPE));
+            }
+
+            docBuilder.translogWriter().writeValue(value);
+        }
+    }
+
+    private static class FixedNumericIndexer extends BaseNumericIndexer {
+
+        private final FieldType fieldType;
+
+        private FixedNumericIndexer(Reference ref, int precision) {
+            super(ref);
+            this.fieldType = new FieldType();
+            this.fieldType.setOmitNorms(true);
+            this.fieldType.setIndexOptions(IndexOptions.DOCS);
+            this.fieldType.setTokenized(false);
+            this.fieldType.setStored(false);
+            this.fieldType.freeze();
+        }
+
+        @Override
+        public void indexValue(@NotNull BigDecimal value, IndexDocumentBuilder docBuilder) throws IOException {
+            BigInteger unscaled = value.unscaledValue();
+            BytesRef binaryValue = new BytesRef(unscaled.toByteArray());
+            if (this.ref.indexType() != IndexType.NONE) {
+                docBuilder.addField(new Field(name, binaryValue, fieldType));
+            }
+            if (ref.hasDocValues()) {
+                docBuilder.addField(new SortedSetDocValuesField(name, binaryValue));
```

Comment:
Using `SortedSetDocValues` gives us multi-valued fields, but given the expected cardinality of these fields it will be inefficient for both storage and retrieval.  IIRC elasticsearch has a multi-valued binary doc values implementation that it might be worth looking at.  Or we could extend ValueIndexer to have an `indexValues()` method that can index an array of values, with a default implementation that just calls `indexValue()` for each individual value, and then specialise it for this case to pack all values together into a single byte array for docvalues storage.

Reply:
Sorry, should have been clearer - I'm not talking about `SortedSetDocValues` vs `SortedDocValues`, I'm talking about `SortedSetDocValues` vs `BinaryDocValues`.  `SortedSet` stores values themselves in a terms dictionary, and then stores the ordinals of those values per document, while `Binary` just stores the bytes per document direclty.  So to look up a `SortedSet` value per document we first load an ordinal, and then lookup its value in the terms dictionary.  This is worth it if lots of documents share values, as we get very good compression; or if the byte array values are wider than a long, it's better for sorting because you can compare the ordinals rather than the values byte-by-byte.  But for very high cardinality fields then you end up storing the value plus a long for every doc, rather than just the value, every value lookup gets an extra indirection, and sorting probably isn't that much more efficient.

Using `SortedSet` is probably fine for now, but something to bear in mind for future improvements.